packages/typegen/readme.md

```diff
@@ -333,6 +333,65 @@ module.exports.default = {
 }
 ```
 
+## Enhancing Return Types
+
+Typgen is designed to output types only to the degree it's certain they are correct.  
+
+Let's say in a complex query it can determine that a specific column will return a `string`, but isn't sure if it is also nullable, it will extract the type as `{ column: string | null }`, just to be on the safe side. When it encounters columns where it is unable to even determine the basic type, i.e. `json` columns, it will return :shrug: (Ok, actually the typescript equivalent, which is `unknown`).
+
+In these cases you likely know more about the actual return type than typegen and you might feel the urge to overwrite the types.  
+Yet you shouldn't touch generated code, as your changes will be removed again on the next run.
+
+Instead what you should do is add (one or more) intersection types to the sql literal, specifying the columns where you want to help typegen out by increasing specificity. The resulting type will be a combination of the extracted types and your enhancements.  
+Check out the [typescript docs on intersection types](https://www.typescriptlang.org/docs/handbook/2/objects.html#intersection-types) to learn more.
+
+Imagine this is your code after running typegen.
+```typescript
+sql<queries.ExtractedResult>`select string_col, json_col from table`
+
+export declare namespace queries {
+  // Generated by @slonik/typegen
+
+  /** - query: `select string_col, json_col from table` */
+  export interface TestTable {
+    /** column: `example_test.table.string_col`, regtype: `character_varying` */
+    string_col: string | null,
+    /** column: `example_test.table.json_col`, regtype: `jsonb` */
+    json_col: unkown
+  }
+}
+```
+
+You can enhance the return type like this:
+
+```typescript
+sql<queries.ExtractedResult & { json_col: string[] }>`[query]`
+```
+\- or, if you prefer -
+```typescript
+interface EnhancedResult {
+  json_col: string[]
+}
+sql<queries.ExtractedResult & EnhancedResult>`[query]`
+```
+
+Either way the resulting type will be this:
+
+```typescript
+type ResultingType = {
+  string_col: string | null,
+  json_col: string[]
+}
+```
+
+**On subsequent runs typegen will only update the first intersection type and leave all following intersections untouched**.
+
+This also means you can make the column `string_col` non-nullable by intersecting it with `{ string_col: string }`.
```

Comment:
I had a thought re the dumb mistake I made that you caught - there's an interesting trick people could do and I can't decide if it's an antipattern or not - you could make a column non-nullable by intersecting with `{ string_col: {} }` instead - then typegen would take care of saying whether it's `string` or `number` or `boolean` - and all the `{}` does, effectively, is exclude `null`. But it's kind of weird and non-obvious, and the resultant type shows in vscode as `string & {}` - which acts like `string` but looks confusing. So I don't know if we should "recommend" it. Curious your thoughts though.

Reply:
Yeah, I share your doubts here. It is very unusual and what I really like about the Intersection solution is that it is very intuitive and readable to users who don't know typegen at all (but have basic understanding of how intersections work). If I saw this code somewhere and something was intersected with `{ col: string }`, I would understand the intention. Where as with `{ string_col: {} }` I would be like "Why is he/she doing that?".
This goes doubly if I don't actually see the intersection but the resulting type being `string & {}` (they might be in different places). I would wonder what this means and follow it to the source.

So in short my recommendation would be to not document, let alone recommend it.

Reply:
Ok - let’s not do it then. Interestingly the typescript team don’t want to introduce a non nullable unknown/any (and it seems I tried to make the case that they should five years ago! https://github.com/microsoft/TypeScript/issues/7648#issuecomment-354036449). If that ever gets picked up then whatever that type were called would be a clearer way to say “typegen can do the hard work, I’ll just say it’s not null”. Anyway no change needed here.